pkg/apis/serving/v1alpha1/service_types.go

```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2019 Kubeflow Community.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ServiceSpec defines the desired state of Service
+type ServiceSpec struct {
+	// If specified, passes name to generated Revision.
+	RevisionName string `json:"rollout,omitempty"`
+	// Must contain either one or two elements.
+	// The first is the "active" revisions and the second is the "candidate" revision.
+	Revisions []string `json:"revisions,omitempty"`
+	// May only be specified if revisions has two elements. Defaults to 0.
+	// This percentage of traffic is routed to the second revision, the remainder to the first.
+	RolloutPercent int `json:"rolloutPercent,omitempty"`
+
+	MinReplicas string `json:"minReplicas,omitempty"`
+	MaxReplicas string `json:"maxReplicas,omitempty"`
+
+	/*
+		The following fields follow a "1-of" semantic. Users must specify exactly one spec.
+	*/
+	Custom      *CustomSpec      `json:"custom,omitempty"`
+	Tensorflow  *TensorflowSpec  `json:"tensorflow,omitempty"`
+	XGBoost     *XGBoostSpec     `json:"xgBoost,omitempty"`
+	ScikitLearn *ScikitLearnSpec `json:"scikitLearn,omitempty"`
+}
+
+// TensorflowSpec defines arguments for configuring Tensorflow model serving.
+type TensorflowSpec struct {
+	ModelUri string `json:"modelUri"`
+	// Defaults to latest TF Version.
+	RuntimeVersion string `json:"runtimeVersion,omitempty"`
+	// Defaults to requests and limits of 1CPU, 2Gb MEM.
+	Resources v1.ResourceRequirements `json:"resources,omitempty"`
+}
+
+// XGBoostSpec defines arguments for configuring XGBoost model serving.
+type XGBoostSpec struct {
+	ModelUri string `json:"modelUri"`
+	// Defaults to latest XGBoost Version.
+	RuntimeVersion string `json:"runtimeVersion,omitempty"`
+	//Defaults to requests and limits of 1CPU, 2Gb MEM.
+	Resources v1.ResourceRequirements `json:"resources,omitempty"`
+}
+
+// ScikitLearnSpec defines arguments for configuring ScikitLearn model serving.
+type ScikitLearnSpec struct {
+	ModelURI string `json:"modelUri"`
+	// Defaults to latest ScikitLearn Version.
+	RuntimeVersion string `json:"runtimeVersion,omitempty"`
+	// Defaults to requests and limits of 1CPU, 2Gb MEM.
+	Resources v1.ResourceRequirements `json:"resources,omitempty"`
+}
+
+// CustomSpec provides a hook for arbitrary container configuration.
+type CustomSpec struct {
+	Container v1.Container `json:"container"`
```

Comment:
Why not use pod instead of a container? Pod is more flexible.

Reply:
I think we may want to retain control over the pod for other purposes.

1) Knative doesn't expose pods
2) If we have any sort of in-pod proxy (e.g. for routing to loggers/explainers/health checks), then the user would need to be aware of these details
3) What specifically, about the pod spec do you think would be beneficial here?

Reply:
This is to do with knative implementation, it only allows specifying one container since the pod launched by knative injects istio-proxy and queue-proxy which communicates to main user container.

Reply:
If we use pod here, we can allow users not to use Knative, to cover more use cases.

If the custom spec is to support other serving frameworks in Knative, then container LGTM